packages/@lwc/engine-core/src/framework/template.ts

```diff
@@ -110,6 +111,68 @@ function validateLightDomTemplate(template: Template, vm: VM) {
     }
 }
 
+// This function serializes the expressions generated by static content optimization.
+// Currently this is only needed for SSR.
+function buildSerializeExpressionFn(parts?: VStaticPart[]) {
+    if (process.env.IS_BROWSER || isUndefined(parts)) {
+        return () => '';
+    }
+    const partIdsToParts = new Map<string, VStaticPart>();
+    for (const staticPart of parts) {
+        partIdsToParts.set(`${staticPart.partId}`, staticPart);
+    }
+    return (partToken: string) => {
+        // The partTokens are split into 3 section:
+        // 1. The first character represents the expression type (attribute, class, style, or text).
+        // 2. The characters from index 1 to the first occurrence of a ':' is the partId.
+        // 3. Everything after the first ':' represents the attribute name.
+        // For example: a0:data-name, a = an attribute, 0 = partId, data-name = attribute name.
+        const type = partToken.charAt(0);
+        const delimiterIndex = partToken.indexOf(':');
+        const partId = partToken.substring(1, delimiterIndex);
+        const rest = partToken.substring(delimiterIndex + 1);
+        const part = partIdsToParts.get(partId) ?? EmptyObject;
```

Comment:
The format of the tokens will be `{type}{partId}:{attributeName}`, for example `a0:data-name`.

The reason this is required is that our current SSR implementation is not compatible with how we are applying attributes in the browser.

The main issue is the `vnode.fragment` used in [`mountStatic`](https://github.com/salesforce/lwc/blob/master/packages/%40lwc/engine-core/src/framework/rendering.ts#L283) has been assembled by this routine during the mount process.

In the browser, we rely on [`cloneNode`](https://github.com/salesforce/lwc/blob/master/packages/%40lwc/engine-core/src/framework/rendering.ts#L291) to create the DOM elements, loop through those elements and apply the attributes to them.

However, in SSR, the [`cloneNode`](https://github.com/salesforce/lwc/blob/master/packages/%40lwc/engine-server/src/renderer.ts#L85-L96) function just returns the string value in `vnode.fragment`. This is why we need to insert the expression values at the time the string is being built.

I thought about replicating the `cloneNode` behavior to reproduce the nodes but it would require an HTML parser to generate the nodes.

I feel that this is the most straight forward approach, although I think we should refactor this function to have different implementations between `@lwc/engine-dom` and `@lwc/engine-server`.

Reply:
Great description! We should capture this in a high-level issue and add a TODO here. I think it would be great to have one place where we document this (very big) difference between `engine-server` and `engine-dom`.

It may ultimately get resolved by the dedicated SSR compiler.

Reply:
Add a issue! #4078